### ISSUE TITLE
Problem using named parameters

### ISSUE THREAD

**Issue by ilsundal** [tedeh/jayson#87]
Hi

I have trouble using named parameters.

I have this minimalist program:

var jayson = require('jayson');
var server = jayson.server({
  add: function(args, callback) {
    console.log(args);
    callback(null, args);
  }
});
server.http().listen(3000);

When I use this curl command:

curl -i POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method":"add", "params": { \"a\": \"hello\" }, "id":1 }' http://localhost:3000

I get this error:

_http_outgoing.js:570
    throw new TypeError('First argument must be a string or Buffer');
    ^

TypeError: First argument must be a string or Buffer
    at ServerResponse.end (_http_outgoing.js:570:11)
    at respond (/home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/lib/utils.js:215:11)
    at /home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/lib/utils.js:188:22
    at /home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/node_modules/lodash/index.js:7173:25
    at /home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/lib/utils.js:157:22
    at Object.Utils.JSON.parse (/home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/lib/utils.js:281:12)
    at IncomingMessage.<anonymous> (/home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/lib/utils.js:156:16)
    at emitNone (events.js:86:13)
    at IncomingMessage.emit (events.js:186:7)
    at endReadableNT (_stream_readable.js:974:12)

However, if I put brackets [ ] and quotes around the params, it works. Using curl:

curl -i POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method":"add", "params": ["{ \"a\": \"hello\" }"], "id":1 }' http://localhost:3000

gives this:

{"jsonrpc":"2.0","id":1,"result":["{ \"a\": \"hello\" }"]}

If I keep an empty object (and no quotes), it works as well. Using curl:

curl -i POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method":"add", "params": { }, "id":1 }' http://localhost:3000

gives this:

{"jsonrpc":"2.0","id":1,"result":{}}

If I put quotes around the object it fails. Using curl:

curl -i POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method":"add", "params": "{ \"a\": \"hello\" }", "id":1 }' http://localhost:3000

gives this:

{"jsonrpc":"2.0","id":null,"error":{"code":-32600,"message":"Invalid request"}}

According to the JSON-RPC v2 spec (http://www.jsonrpc.org/specification), I should be able to use named parameters without quotes, e.g.

--> {"jsonrpc": "2.0", "method": "subtract", "params": {"subtrahend": 23, "minuend": 42}, "id": 3}
<-- {"jsonrpc": "2.0", "result": 19, "id": 3}

I have used a JSON linter (http://jsonlint.com/) as well. And the named parameters without brackets and quotes works there as well.

However, I cannot get it to work with Jayson. What am I doing wrong?

Thanks!

**Comment by tedeh** [comment]
I tested your server and curl command against the latest commit in the master branch (bc99be626eaf45adb31a1c0f660da1d5bcc2490e at the time of writing) and it worked fine, so I'm not sure what is wrong.

I also removed the `-i POST` flag which is not necessary and not used for specifying the method, see the `-X` flag if you want to do that).

**Comment by ilsundal** [comment]
If I run this command:

curl -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method":"add", "params": "{ \"a\": \"hello\" }", "id":1 }' http://localhost:3000

I get this response:

{"jsonrpc":"2.0","id":null,"error":{"code":-32600,"message":"Invalid request"}}

Why is the request invalid? I am escaping the quotes wrongly, or..?

**Comment by tedeh** [comment]
You should not put quotes around the "params" object. This is not a problem with jayson, closing.

**Comment by ilsundal** [comment]
Well, if I omit those quotes like this:

curl -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method":"add", "params": { \"a\": \"hello\" }, "id":1 }' http://localhost:3000

then I get this Jayson error:

_http_outgoing.js:570
    throw new TypeError('First argument must be a string or Buffer');
    ^

TypeError: First argument must be a string or Buffer
    at ServerResponse.end (_http_outgoing.js:570:11)
    at respond (/home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/lib/utils.js:215:11)
    at /home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/lib/utils.js:188:22
    at /home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/node_modules/lodash/index.js:7173:25
    at /home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/lib/utils.js:157:22
    at Object.Utils.JSON.parse (/home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/lib/utils.js:281:12)
    at IncomingMessage.<anonymous> (/home/morten/NewBanking/repos/nbapiserver/node_modules/jayson/lib/utils.js:156:16)
    at emitNone (events.js:86:13)
    at IncomingMessage.emit (events.js:186:7)
    at endReadableNT (_stream_readable.js:974:12)

as described in my original report.

**Comment by tedeh** [comment]
So I discovered there was a bug in the jayson http server which caused a crash when given wacky JSON. This was fixed in the referenced commit and published in `v2.0.3`. Generally you are recommended to use the middleware instead, the http server is mostly for reference because you can't control exactly what happens in cases like this.

An attempt with valid JSON in `v2.0.3`:
```
$ curl -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method":"add", "params": { "a": 23 }, "id":1 }' http://localhost:3000
{ a: 23 }
```

An attempt with invalid JSON:
```
$ curl -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method":"add", "params": "{ "a": "hello" }", "id":1 }' http://localhost:3000
SyntaxError: Unexpected token a in JSON at position 49
```
